treefile: Add new `add-commit-metadata` key

jlebon opened the pull request:
Add support for a new `add-commit-metadata` key in the treefile so that
we can directly specify commit metadata we want to inject from there.

This will be useful in Fedora CoreOS, where we'll have separate
treefiles for each streams, each with stream-specific metadata values
required.

jlebon commented:
See https://github.com/coreos/fedora-coreos-config/pull/110 for example usage.

lucab commented on rust/src/treefile.rs:
```diff
@@ -279,6 +279,22 @@ fn merge_vec_field<T>(dest: &mut Option<Vec<T>>, src: &mut Option<Vec<T>>) {
+    if let Some(mut srcv) = src.take() {
+        if let Some(mut destv) = dest.take() {
+            for (k, v) in destv.drain() {
+                srcv.insert(k, v);
```
I think you are manually rolling the equivalent of https://doc.rust-lang.org/std/iter/trait.Extend.html#tymethod.extend.

jlebon replied:
Heh, it's funny. I'm working on another rpm-ostree patch where I did find `.extend()` and use it. But somehow missed it would've been useful in this patch as well. Thanks, will fix!

jlebon commented:
Hmm OK and the Rust compile errors are the same as the ones in https://github.com/projectatomic/rpm-ostree/pull/1863#issuecomment-508550414. Will dig into this.

jlebon commented:
So the issue was in fact that we didn't track `Cargo.lock` for bindgen. Took me a while to figure out it wasn't the main `librpmostree_rust` library that was failing to build (even though looking closely at the logfile now, it seems pretty obvious).

Anyway, as mentioned in the commit message, we should be able to update to 1.35.0 soon too.

rfairley commented on src/app/rpmostree-compose-builtin-tree.c:
```diff
@@ -713,6 +703,41 @@ rpm_ostree_compose_context_new (const char    *treefile_pathstr,
+        return FALSE;
+    }
+
+
```
Minor: stray newline

jlebon commented:
> ```
> File 'out.txt' doesn't match fixed string list 'No apparent changes since previous commit'
> ```

Hmm, I'm not hitting this locally. Still trying to figure out what's going on here.

rfairley commented:
Also not seeing this locally, [running the CI command](https://github.com/projectatomic/rpm-ostree/blob/95d1c6ea29c8d40bc4eab211ada17454bb8ad712/.papr.yml#L138) . CI is also showing

```
error: No such metadata key 'exampleos.gitrepo'
```

in `f29-compose1`.

jlebon commented:
Yeah, the issue is that we'll get different input hashes based on how serde decides to serialize the hash table (and so the ordering of keys can change on each iteration). I'm working on adding checksumming over all the file inputs to fix this (incidentally also fixing: https://github.com/projectatomic/rpm-ostree/blob/c94bd08b02499a3f2a5804f5c2ba207a0afbda18/src/app/rpmostree-composeutil.c#L90).

jlebon commented:
Added docs and fixed tests! :arrow_up: 

> and so the ordering of keys can change on each iteration

In fact, the only change necessary for this was to swap out `HashMap` for a `BTreeMap`, since those keys are ordered. But I ended up also enhancing the way we hash inputs anyway in the process. We can split out that last commit into a separate patch if wanted. It's not strictly needed for FCOS.

lucab commented:
> Notice we hash the *reserialization* of the final flattened treefile only so that e.g.
> comments/whitespace/hash table key reorderings don't trigger a respin.

This is an high-level pattern which always causes me troubles, as it is hashing some input state that doesn't really exist and is not persisted. Would you consider just hashing the raw input bytes instead?

jlebon commented:
> Would you consider just hashing the raw input bytes instead?

Yeah, that's in fact what I initially had. I then went back to just covering the final flattened treefile (which is the status quo) when I realized it regressed on detecting what is essentially no-ops. Is the argument for it in that quote not motivation enough in your opinion?

> as it is hashing some input state that doesn't really exist and is not persisted.

This is not strictly true; the flattened treefile is written to `/usr/share/rpm-ostree/treefile.json` in the OSTree. :) So in theory, we could still retrieve the treefile hash for a previous commit if needed.

ISTM like hashing all the treefiles is more of an escape hatch for being unable to easily deal with some tricky situations. Seems like it's worth trying the "smart approach" first before resorting to the easier one?

That said, I'm not strongly against it either. Just slightly more leaning towards maintaining the status quo until there's an obvious problem with that (or we discover some subtleties/interactions with how this behaviour integrates with e.g. COSA or the FCOS/RHCOS pipelines in a negative way).

lucab commented:
I don't grasp the full context of the effects of those no-ops, so I am up to your call here.

cgwalters commented:
Another approach is to hash the [JSON canonical form](https://github.com/zmanian/canonical_json) - also to write that by default.

cgwalters commented on rust/src/treefile.rs:
```diff
@@ -516,6 +524,33 @@ for x in *; do mv ${{x}} %{{buildroot}}%{{_prefix}}/lib/ostree-jigdo/%{{name}};
+fn hash_file(hasher: &mut glib::Checksum, mut f: &fs::File) -> Fallible<()> {
+    let mut reader = io::BufReader::new(f.try_clone()?);
+    let mut buf = Vec::new();
+    reader.read_to_end(&mut buf)?;
```
Doesn't matter a lot here but probably worth a comment like
`// If we were to use this for large files we should read into a e.g. 128k buffer`

jlebon replied:
For fun, I switched to chunked reads below! :arrow_down:

cgwalters commented:
Any time we change the checksum algorithm, it will trigger a "pointless" build for users.  This isn't a *serious* problem IMO, but we shouldn't do it often.  Concretely for example, after this lands in Fedora, depending on whether or not there are AH/Silverblue updates pending, users may get a "no-op" update.   

IOW, if we're going to change it now we should feel it's pretty future proof.

I like the direction the code is going in.  However, it looks like we're relying on the Serde serialization, and I suspect that's not defined to be stable - they would feel free to e.g. add whitespace?  It looks like that canonical JSON library I linked above is abandoned though :frowning: 

I dunno.  Maybe it's unlikely Serde will change and we can just live with that.

jlebon commented:
> Any time we change the checksum algorithm, it will trigger a "pointless" build for users. This isn't a serious problem IMO, but we shouldn't do it often.

Yeah, agreed on both counts.

>  It looks like that canonical JSON library I linked above is abandoned though

Ohh, that would've been perfect for us. I might create an RFE for serde_json to add a canonical mode to see what they think. I guess one thing we could do at least in the mean time is checksum in non-pretty mode.

WRT checksumming all treefiles vs just the final flattened one, I think it comes down to which one will be the most "unstable", serde's format vs humans/robots making inconsequential changes to treefiles... I'd think it'd be more of the latter :)

Note that the way we checksum things now is purely additive (i.e. we were already checksumming the flattened treefile before; we're just adding more things to the checksum). So at the very least, it's not going to be *less* stable than it already was.

cgwalters commented:
>  I guess one thing we could do at least in the mean time is checksum in non-pretty mode.

Oh this definitely; the pretty mode seems much more likely to change.

> So at the very least, it's not going to be less stable than it already was.

Right; as you noted in the commit message the original core issue was using `HashMap` instead of `BTreeMap` which you fixed.

Anyways...from my PoV we don't need to keep rehashing this more (so to speak), I'm good with the PR with the addition of using non-pretty mode per above.

jlebon commented:
OK updated! I also added to the commit message:

    One notable change though is that we now hash the treefile in non-pretty
    mode to increase the chances that the serialized form remains stable.
    Ironically, this change is likely to cause a no-op commit once it gets
    to pipelines which iterate quickly. All for the greater good though.

:smile:

cgwalters commented:
@rh-atomic-bot r+ 8b3753f

jlebon commented:
```
+ coreos-assembler init --force https://github.com/coreos/fedora-coreos-config
stat: cannot stat '/dev/kvm': No such file or directory
```

Hmm, were we scheduled on a node that doesn't have nested virt or something? This definitely works usually.

@rh-atomic-bot retry

lucab commented:
Relevant to the "serde canonical json" conversation: https://github.com/serde-rs/json/issues/309
